### PR TITLE
EP-705 iOS - Make a server payment

### PR DESCRIPTION
**Pull request by AlexeiJovmir**
return response if type is server to server for default and apple type.

### CODE REVIEW

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m [hunk 1]
```diff
@@ -221,4 +229,42 @@ - (JP3DSService *)threeDSecureService {
...
+    data.paymentReference = self.configuration.reference.paymentReference;
+    
+    NSDateFormatter* df = [[NSDateFormatter alloc] init];
+    [df setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSSSZZZZZ"];
+    NSString *result = [df stringFromDate:[NSDate date]];
+    data.createdAt = result;
+    
+    data.consumer = [[JPConsumer alloc] init];
+    data.consumer.consumerReference = self.configuration.reference.consumerReference;
+    data.consumer.consumerToken = self.configuration.reference.paymentReference;
```
Comment (marcghelas) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m:254]: I don't think consumerToken is the same as paymentReference
better map to https://github.com/Judopay/JudoKitObjC/blob/develop-v2/Source/Models/TransactionData/JPTransactionData.h#L43
as on android

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m [hunk 2]
```diff
@@ -221,4 +229,42 @@ - (JP3DSService *)threeDSecureService {
...
+    data.amount = self.configuration.amount;
+    data.cardDetails = [[JPCardDetails alloc] init];
+    JPStoredCardDetails *selectedCard = [self selectedCard];
+    data.cardDetails.cardLastFour = selectedCard.cardLastFour;
+    data.cardDetails.cardToken = selectedCard.cardToken;
+    data.cardDetails.cardNetwork = selectedCard.cardNetwork;
```
Comment (marcghelas) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m:260]: On android we also map cardDetails.scheme
https://github.com/Judopay/JudoKitObjC/blob/develop-v2/Source/Models/CardDetails/JPCardDetails.h#L155
which is network name (Visa, Mastercard etc.)

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m (hunk 1 above)
Comment (AlexeiJovmir) [reply Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m:254]: on andoird side it is optional, so i will remove it from ios

File: Source/Services/ApplePay/JPApplePayService.m [hunk 3]
```diff
@@ -320,4 +328,33 @@ - (nullable JPContactInformation *)contactInformationFromPaymentContact:(nullabl
                                                 postalAddress:postalAddress];
 }
 
+- (void)processServerToServer:(JudoCompletionBlock)completion payment:(PKPayment *)payment {
+    completion([self buildResponse: payment], nil);
+}
+
+- (JPResponse *)buildResponse:(PKPayment *)payment  {
+    //proces payment in response
```
Comment (mpetrenco) [inline Source/Services/ApplePay/JPApplePayService.m:336]: Consider removing the comment

File: Source/Services/ApplePay/JPApplePayService.m [hunk 4]
```diff
@@ -320,4 +328,33 @@ - (nullable JPContactInformation *)contactInformationFromPaymentContact:(nullabl
...
+    JPResponse *response = [[JPResponse alloc] initWithPagination:nil];
+    
+    JPTransactionData *data = [[JPTransactionData alloc] init];
```
Comment (mpetrenco) [inline Source/Services/ApplePay/JPApplePayService.m:339]: [JPTransactionData new]

File: Source/Services/ApplePay/JPApplePayService.m [hunk 5]
```diff
@@ -320,4 +328,33 @@ - (nullable JPContactInformation *)contactInformationFromPaymentContact:(nullabl
...
+    data.judoId = self.configuration.judoId;
+    data.paymentReference = self.configuration.reference.paymentReference;
+    
+    NSDateFormatter* df = [[NSDateFormatter alloc] init];
```
Comment (mpetrenco) [inline Source/Services/ApplePay/JPApplePayService.m:343]: [NSDateFormatter new]

File: Source/Services/ApplePay/JPApplePayService.m [hunk 6]
```diff
@@ -320,4 +328,33 @@ - (nullable JPContactInformation *)contactInformationFromPaymentContact:(nullabl
...
+    [df setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSSSZZZZZ"];
+    NSString *result = [df stringFromDate:[NSDate date]];
+    data.createdAt = result;
+    
+    data.consumer = [[JPConsumer alloc] init];
```
Comment (mpetrenco) [inline Source/Services/ApplePay/JPApplePayService.m:348]: [JPConsumer new]

File: Source/Services/ApplePay/JPApplePayService.m [hunk 7]
```diff
@@ -320,4 +328,33 @@ - (nullable JPContactInformation *)contactInformationFromPaymentContact:(nullabl
...
+    data.consumer.consumerReference = self.configuration.reference.consumerReference;
+    data.consumer.consumerToken = self.configuration.reference.paymentReference;
```
Comment (mpetrenco) [inline Source/Services/ApplePay/JPApplePayService.m:350]: Is this similar to Android?

Comment (AlexeiJovmir) [reply Source/Services/ApplePay/JPApplePayService.m:350]: removed from JPPaymentMethodsInteractorImpl.. will remove from apple service as well

File: Source/Services/ApplePay/JPApplePayService.m [hunk 8]
```diff
@@ -320,4 +328,33 @@ - (nullable JPContactInformation *)contactInformationFromPaymentContact:(nullabl
...
+    data.amount = self.configuration.amount;
+    data.cardDetails = [[JPCardDetails alloc] init];
+    
+    data.cardDetails.cardToken = payment.token.transactionIdentifier;
```
Comment (mpetrenco) [inline Source/Services/ApplePay/JPApplePayService.m:354]: I don't think the transaction identifier is the same as card token

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m [hunk 9]
```diff
@@ -221,4 +229,42 @@ - (JP3DSService *)threeDSecureService {
...
+        if (card.isSelected) {
+            return card;
+        }
+    }
+    return nil;
+}
+
+- (JPResponse *)buildResponse {
+    JPResponse *response = [[JPResponse alloc] initWithPagination:nil];
+    JPTransactionData *data = [[JPTransactionData alloc] init];
```
Comment (mpetrenco) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m]: [JPTransactionData new];

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m [hunk 10]
```diff
@@ -221,4 +229,42 @@ - (JP3DSService *)threeDSecureService {
...
+    data.judoId = self.configuration.judoId;
+    data.paymentReference = self.configuration.reference.paymentReference;
+    
+    NSDateFormatter* df = [[NSDateFormatter alloc] init];
```
Comment (mpetrenco) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m]: Let's set the date formatter as a lazy variable and reuse it.

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m [hunk 11]
```diff
@@ -221,4 +229,42 @@ - (JP3DSService *)threeDSecureService {
...
+    [df setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSSSZZZZZ"];
+    NSString *result = [df stringFromDate:[NSDate date]];
+    data.createdAt = result;
+    
+    data.consumer = [[JPConsumer alloc] init];
```
Comment (mpetrenco) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m]: [JPConsumer new]

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m [hunk 12]
```diff
@@ -221,4 +229,42 @@ - (JP3DSService *)threeDSecureService {
...
+    data.consumer.consumerReference = self.configuration.reference.consumerReference;
+    data.consumer.consumerToken = self.configuration.reference.paymentReference;
+    data.amount = self.configuration.amount;
+    data.cardDetails = [[JPCardDetails alloc] init];
```
Comment (mpetrenco) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m]: [JPCardDetails new]

File: JudoKitSwiftExampleApp/ViewControllers/MainViewController.swift [hunk 13]
```diff
@@ -211,6 +218,15 @@ class MainViewController: UITableViewController {
         })
     }
     
+    @objc func navigateToServerToServerMethods() {
```
Comment (stpopa) [inline JudoKitSwiftExampleApp/ViewControllers/MainViewController.swift:221]: the main sample app - the one we distribute, is the Objc sample, please update that thing first, the Swift one should be kept in sync, but it's not a must

File: Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m [hunk 14]
```diff
@@ -221,4 +230,48 @@ - (JP3DSService *)threeDSecureService {
...
+    return nil;
+}
+
+- (JPResponse *)buildResponse {
+    JPResponse *response = [JPResponse new];
+    JPTransactionData *data = [JPTransactionData new];
+    data.judoId = self.configuration.judoId;
+    data.paymentReference = self.configuration.reference.paymentReference;
+    
+    NSString *result = [self.dateFormatter stringFromDate:[NSDate date]];
```
Comment (stpopa) [inline Source/Modules/PaymentMethods/Interactor/JPPaymentMethodsInteractor.m:248]: `[NSDate date]` -> `NSDate.date`

File: Source/Services/ApplePay/JPApplePayService.m [hunk 15]
```diff
@@ -320,4 +329,37 @@ - (nullable JPContactInformation *)contactInformationFromPaymentContact:(nullabl
...
+}
+
+- (JPResponse *)buildResponse:(PKPayment *)payment  {
+    JPResponse *response = [JPResponse new];
+    
+    JPTransactionData *data = [JPTransactionData new];
+    data.judoId = self.configuration.judoId;
+    data.paymentReference = self.configuration.reference.paymentReference;
+    
+    NSString *result = [self.dateFormatter stringFromDate:[NSDate date]];
```
Comment (stpopa) [inline Source/Services/ApplePay/JPApplePayService.m:343]: [NSDate date] -> NSDate.date

File: Source/Services/ApplePay/JPApplePayService.m [hunk 16]
```diff
@@ -320,4 +329,37 @@ - (nullable JPContactInformation *)contactInformationFromPaymentContact:(nullabl
...
+    data.amount = self.configuration.amount;
+    data.cardDetails = [JPCardDetails new];
+    
+    data.cardDetails.cardToken = payment.token.transactionIdentifier;
+    data.cardDetails.cardScheme = payment.token.paymentMethod.network;
+    response.items = @[data];
+    return response;
+}
+
+- (NSDateFormatter *)dateFormatter {
```
Comment (stpopa) [inline Source/Services/ApplePay/JPApplePayService.m:357]: duplicated function from `JPPaymentMethodsInteractor.m:269`, let's extract this to a singleton class

Comment (stpopa) [reply Source/Services/ApplePay/JPApplePayService.m:357]:
```
@implementation JPFormatters {

+ (JPFormatters *)sharedInstance {
    static JPFormatters *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [JPFormatters alloc new];
    });

     return sharedInstance;
}

- (NSDateFormatter *)dateFormatter {
    if (_dateFormatter == nil) {
        _dateFormatter = [[NSDateFormatter alloc] init];
        [_dateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss.SSSSZZZZZ"];
    }
    return _dateFormatter;
}

}
```

**Pull request closed without merging**